### PR TITLE
DOCSP-42325: Bulk Write

### PR DESCRIPTION
**Pull request by mcmorisi**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-scala/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-42325>
Staging - <https://deploy-preview-68--docs-scala.netlify.app/write/bulk-write/>

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/write/bulk-write.txt [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+Additional Information
+----------------------
+
+To learn how to perform individual write operations, see the following guides:
+
+- :ref:`scala-write-insert`
+- :ref:`scala-write-update`
+- :ref:`scala-write-delete`
+- :ref:`scala-write-replace`
```
Comment (mcmorisi) [inline source/write/bulk-write.txt:327]: Broken links that should be resolved when all the pages are merged

File: source/write/bulk-write.txt [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+   ``MongoBulkWriteException``.
+
+Update Operations
+~~~~~~~~~~~~~~~~~
+
+To update a document, create an instance of ``UpdateOneModel`` and pass
+the following arguments:
+
+- A **query filter** that specifies the criteria used to match documents in your collection
+- The update operation you want to perform. For more information about update
```
Comment (norareidy) [inline source/write/bulk-write.txt:103]: S: Remove the articles at the beginning of each  [list item](https://www.mongodb.com/docs/meta/style-guide/style/lists/#list-items); also, since the second bullet point has periods I'd add one to the first:

```suggestion
- **Query filter** that specifies the criteria used to match documents in your collection.
- Update operation you want to perform. For more information about update
```

File: source/write/bulk-write.txt [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+   :dedent:
+
+Replace Operations
+~~~~~~~~~~~~~~~~~~
+
+A replace operation removes all fields and values of a specified document and
+replaces them with new fields and values that you specify. To perform a
+replace operation, create an instance of ``ReplaceOneModel`` and pass a
+query filter and the fields and values you want to replace the matching
+document with.
```
Comment (norareidy) [inline source/write/bulk-write.txt:140]: S: not necessary, but you can list the arguments here like you did with the update section:
```suggestion
replace operation, create an instance of ``ReplaceOneModel`` and pass the following arguments:

- Query filter ...
- Replacement document ...
```

File: source/includes/write/bulk-write.scala [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                                "cuisine" -> "American")
+      val replaceOneModel = ReplaceOneModel(replaceFilter, replaceDoc)
+      // end-bulk-replace-one
+
+      // start-bulk-delete-one
+      val deleteOneModel = DeleteOneModel(equal("name", "Morris Park Bake Shop"))
+      // end-bulk-delete-one
+
+      // start-bulk-delete-many
+      val deleteMany = DeleteManyModel(equal("cuisine", "Experimental"))
```
Comment (norareidy) [inline source/includes/write/bulk-write.scala:51]: S: [nit] use the same variable naming pattern as the other examples
```suggestion
      val deleteManyModel = DeleteManyModel(equal("cuisine", "Experimental"))
```

File: source/write/bulk-write.txt [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+
+   .. input:: /includes/write/bulk-write.scala
+      :start-after: start-bulk-write-mixed
+      :end-before: end-bulk-write-mixed
+      :language: scala
+      :dedent:
+
+   .. output::
+      
+      Success: AcknowledgedBulkWriteResult{insertedCount=1, matchedCount=1, removedCount=1, modifiedCount=1, upserts=[], inserts=[BulkWriteInsert{index=0, id=BsonObjectId{value=...}}]}
```
Comment (norareidy) [inline source/write/bulk-write.txt:208]: S: you can add some newlines here to reduce scroll
```suggestion
      Success: AcknowledgedBulkWriteResult{insertedCount=1, matchedCount=1, 
      removedCount=1, modifiedCount=1, upserts=[], inserts=
      [BulkWriteInsert{index=0, id=BsonObjectId{value=...}}]}
```

File: source/write/bulk-write.txt [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+      Completed
+
+If any of the write operations fail, the {+driver-short+} raises a
+``BulkWriteError`` and does not perform any further operations.
+``BulkWriteError`` provides a ``details`` item that includes the
+operation that failed, and details about the exception.
```
Comment (norareidy) [inline source/write/bulk-write.txt:214]: S: remove the comma before the dependent clause
```suggestion
operation that failed and details about the exception.
```

File: source/write/bulk-write.txt [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+- :ref:`scala-write-insert`
+- :ref:`scala-write-update`
+- :ref:`scala-write-delete`
+- :ref:`scala-write-replace`
+
+API Documentation
+~~~~~~~~~~~~~~~~~
+
+To learn more about any of the methods or types discussed in this
+guide, see the following API Documentation:
```
Comment (norareidy) [inline source/write/bulk-write.txt:333]: S:
```suggestion
guide, see the following API documentation:
```

File: source/write/bulk-write.txt [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+
+   * - ``comment()``
+     - | Sets a comment to attach to the operation.
+
+   * - ``let()``
+     - | Provides a map of parameter names and values to set top-level
+         variables for the operation. Values must be constant or closed
+         expressions that don't reference document fields.
+
+The following code creates options and uses the ``ordered(false)`` option to
```
Comment (norareidy) [inline source/write/bulk-write.txt:265]: S: I think this wording makes slightly more sense, since `ordered(false)` isn't the option:
```suggestion
The following code creates options and sets the ``ordered`` option to ``false`` to
```

File: source/write/bulk-write.txt [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+.. note::
+
+   Unordered bulk operations do not guarantee an order of execution. The
+   order can differ from the way you list them to optimize the runtime.
+
+Return Value
+------------
+
+The ``bulkWrite()`` method returns a ``SingleObservable`` object. You can subscribe to
+this observable to access information from a ``BulkWriteResult`` instance by using the
```
Comment (norareidy) [inline source/write/bulk-write.txt:288]: S: this wording is a little confusing, since it's unclear if the method returns a SingleObservable or a BulkWriteResult. From the API docs, it looks like it returns a SingleObservable that contains a BulkWriteResult. Maybe reword to: 

```suggestion
The ``bulkWrite()`` method returns a ``SingleObservable`` object that contains a ``BulkWriteResult``. You can access information from the ``BulkWriteResult`` instance by subscribing to the observable and using the
```

**Review by norareidy — CHANGES_REQUESTED** [review]
Nice work, I left some (mostly minor) suggestions!

File: source/write/bulk-write.txt [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+.. note::
+
+   When the driver runs a bulk operation, it uses the write concern of the
+   target collection. The driver reports all write concern errors after
+   attempting all operations, regardless of execution order.
+
+Customize Bulk Write Operation
+------------------------------
+
+The ``bulkWrite()`` method optionally accepts a parameter which
```
Comment (norareidy) [inline source/write/bulk-write.txt:228]: S: you can introduce BulkWriteOptions here, since you mention it in the next paragraph
```suggestion
The ``bulkWrite()`` method optionally accepts a ``BulkWriteOptions`` parameter, which
```

File: source/write/bulk-write.txt [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+After you define a model instance for each operation you want to perform,
+pass a list of these instances to the ``bulkWrite()`` method.
+By default, the method runs the operations in the order
+specified by the list of models.
+
+The following example performs multiple write operations by using the
+``bulkWrite()`` method:
+
+.. io-code-block::
+
```
Comment (norareidy) [inline source/write/bulk-write.txt:201]: S: make this example copyable:
```suggestion
.. io-code-block::
  :copyable:
```

File: source/write/bulk-write.txt [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+   .. input:: /includes/write/bulk-write.scala
+      :start-after: start-bulk-write-mixed
+      :end-before: end-bulk-write-mixed
+      :language: scala
+      :dedent:
+
+   .. output::
+
```
Comment (norareidy) [inline source/write/bulk-write.txt:209]: S:
```suggestion
        :visible: false
```

**Review by norareidy — APPROVED** [review]
LGTM + some small suggestions

**Review by jyemin — APPROVED** [review]
